net/protocol/replicator.go

```diff
@@ -0,0 +1,146 @@
+	replicatorProtocolResponse = "/defradb/rep_resp/0.0.1"
+)
+
+type pushLogFunc func(
+	ctx context.Context,
+	req *PushLogRequest,
+	isReplicator bool,
+) (*PushLogReply, error)
+
+type replicatorFailureFunc func(ctx context.Context, peerID, docID string) error
+
+// ReplicatorProtocol is the protocol implementation for sending resource updates to a peer node.
+type ReplicatorProtocol struct {
+	*baseProto
+	pushLogFunc           pushLogFunc
+	replicatorFailureFunc replicatorFailureFunc
+}
+
+// NewReplicatorProtocol returns and a new [ReplicatorProtocol] struct and registers the protocol
+// on the stream handler.
+func NewReplicatorProtocol(
+	h host.Host,
+	pushLogFunc pushLogFunc,
+	replicatorFailureFunc replicatorFailureFunc,
+) *ReplicatorProtocol {
+	proto := &ReplicatorProtocol{
+		baseProto:             newBaseProto(h),
+		pushLogFunc:           pushLogFunc,
+		replicatorFailureFunc: replicatorFailureFunc,
+	}
+	h.SetStreamHandler(replicatorProtocolRequest, proto.onRequest)
+	h.SetStreamHandler(replicatorProtocolResponse, proto.onResponse)
+	return proto
+}
+
+// PushToReplicator sends the pushlog request to the provided peer node.
+//
+// Callers should set an appropriate context timeout.
+func (proto *ReplicatorProtocol) PushToReplicator(
+	ctx context.Context,
+	evt event.Update,
+	pid peer.ID,
+) (reply *PushLogReply, err error) {
+	defer func() {
+		// When the event is a retry, we don't need to republish the failure as
+		// it is already being handled by the retry mechanism through the success channel.
+		if err != nil && !evt.IsRetry {
+			handleRepErr := proto.replicatorFailureFunc(ctx, pid.String(), evt.DocID)
+			if handleRepErr != nil {
+				err = errors.Join(err, handleRepErr)
+			}
+		}
+	}()
+
+	req := PushLogRequest{
+		DocID:        evt.DocID,
+		CID:          evt.Cid.Bytes(),
+		CollectionID: evt.CollectionID,
+		Creator:      proto.host.ID().String(),
+		Block:        evt.Block,
+	}
+	m, err := message.Send(ctx, proto, &req, pid, replicatorProtocolRequest, true)
+	if err != nil {
+		return nil, err
+	}
+	return m.(*PushLogReply), nil //nolint:forcetypeassert
+}
+
+func (proto *ReplicatorProtocol) onRequest(s network.Stream) {
+	ctx := context.Background()
+	var err error
+
+	req := PushLogRequest{}
+	err = message.Receive(s, proto, &req)
+	if err != nil {
+		return
+	}
+
+	defer func() {
+		// if an error occurs, try to tell the node that sent the request what went wrong.
+		if err != nil {
+			resp := PushLogReply{}
+			resp.SetMessageID(req.MessageID)
+			resp.SetErrMessage(err.Error())
+			_, _ = message.Send(ctx, proto, &resp, s.Conn().RemotePeer(), replicatorProtocolResponse, false)
+		}
+	}()
+
+	resp, err := proto.pushLogFunc(ctx, &req, true)
+	if err != nil {
+		return
+	}
+
+	resp.SetMessageID(req.MessageID)
+	_, _ = message.Send(ctx, proto, resp, s.Conn().RemotePeer(), replicatorProtocolResponse, false)
+}
+
+func (proto *ReplicatorProtocol) onResponse(s network.Stream) {
```

Comment:
question: if the responses are not used could they instead be an optional part of the interface?